parmed/modeller/residue.py

```diff
@@ -329,6 +377,100 @@ def fix_charges(self, to=None, precision=4):
+
+        An exception is thrown if patch is incompatible because
+        * The patch specifies that an atom is to be deleted that doesn't exist in the residue
+        * A bond specified as being added in the patch does not have both atom names present after adding/deleting atoms from the patch
+        * The new net charge is not integral to the specified precision
+
+        Parameters
+        ----------
+
+        patch : PatchTemplate
+            The patch to apply to this residue
+
+        precision : int, optional
+            Each valid patch should be produce a net charge that is integral to
+            this many decimal places.
+            Default is 4
+
+        Returns
+        -------
+
+        residue : ResidueTemplate
+            A new ResidueTemplate corresponding to the patched residue is returned.
+            The original remains unmodified.
+
+        """
+        # Create a copy
+        # TODO: Once ResidueTemplate.from_residue() actually copies all info, use that instead?
+        residue = _copy.copy(self)
+        # Delete atoms
+        for atom_name in patch.delete_atoms:
+            residue.delete_atom(atom_name)
+        # Add or replace atoms
+        for atom in patch.atoms:
+            if atom.name in residue:
+                # Overwrite type and charge
+                residue[atom.name].type = atom.type
+                residue[atom.name].charge = atom.charge
+            else:
+                residue.add_atom(Atom(name=atom.name, type=atom.type, charge=atom.charge))
+        # Add bonds
+        for (atom1_name, atom2_name, order) in patch.add_bonds:
+            try:
+                # Remove dangling bonds
+                for name in [atom1_name, atom2_name]:
+                    if name == residue.head:
+                        residue.head = None
+                    if name == residue.tail:
+                        residue.tail = None
+                # Add bond
+                residue.add_bond(atom1_name, atom2_name, order)
+            except:
+                raise Exception('Bond %s-%s could not be added to patched residue: atoms are %s' % (atom1_name, atom2_name, list(residue._map.keys())))
+        # Delete impropers
+        for impr in patch.delete_impropers:
+            try:
+                residue.remove(impr)
+            except:
+                raise Exception('Improper %s was not found in residue to be patched.' % impr)
+        # Check that the net charge is integral.
+        net_charge = residue.net_charge
+        is_integral = (round(net_charge, precision) - round(net_charge)) == 0.0
+        if not is_integral:
+            raise Exception('Patch is not compatible with residue due to non-integral charge (charge was %f).' % net_charge)
+        # Ensure residue is connected
+        import networkx as nx
+        G = residue.to_networkx()
+        if not nx.is_connected(G):
+            print([ c for c in nx.connected_components(G) ])
```

Comment:
Either remove the print statement or use the `logging` module and print this at the debug level with some more information.

Reply:
Incorporating this debug information into the exception.